google-ads-examples/src/main/java/com/google/ads/googleads/examples/advancedoperations/AddDynamicPageFeed.java

```diff
@@ -0,0 +1,540 @@
+import com.google.ads.googleads.v1.resources.FeedAttribute;
+import com.google.ads.googleads.v1.resources.FeedItem;
+import com.google.ads.googleads.v1.resources.FeedItemAttributeValue;
+import com.google.ads.googleads.v1.resources.FeedMapping;
+import com.google.ads.googleads.v1.services.AdGroupCriterionOperation;
+import com.google.ads.googleads.v1.services.AdGroupCriterionServiceClient;
+import com.google.ads.googleads.v1.services.CampaignOperation;
+import com.google.ads.googleads.v1.services.CampaignServiceClient;
+import com.google.ads.googleads.v1.services.FeedItemOperation;
+import com.google.ads.googleads.v1.services.FeedItemServiceClient;
+import com.google.ads.googleads.v1.services.FeedMappingOperation;
+import com.google.ads.googleads.v1.services.FeedMappingServiceClient;
+import com.google.ads.googleads.v1.services.FeedOperation;
+import com.google.ads.googleads.v1.services.FeedServiceClient;
+import com.google.ads.googleads.v1.services.GoogleAdsRow;
+import com.google.ads.googleads.v1.services.GoogleAdsServiceClient;
+import com.google.ads.googleads.v1.services.GoogleAdsServiceClient.SearchPagedResponse;
+import com.google.ads.googleads.v1.services.MutateAdGroupCriteriaResponse;
+import com.google.ads.googleads.v1.services.MutateAdGroupCriterionResult;
+import com.google.ads.googleads.v1.services.MutateCampaignResult;
+import com.google.ads.googleads.v1.services.MutateCampaignsResponse;
+import com.google.ads.googleads.v1.services.MutateFeedItemResult;
+import com.google.ads.googleads.v1.services.MutateFeedItemsResponse;
+import com.google.ads.googleads.v1.services.MutateFeedMappingResult;
+import com.google.ads.googleads.v1.services.MutateFeedMappingsResponse;
+import com.google.ads.googleads.v1.services.MutateFeedsResponse;
+import com.google.ads.googleads.v1.services.SearchGoogleAdsRequest;
+import com.google.ads.googleads.v1.utils.ResourceNames;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.Int64Value;
+import com.google.protobuf.StringValue;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.*;
+
+/**
+ * This code example adds a page feed to specify precisely which URLs to use with your Dynamic
+ * Search Ads campaign.
+ */
+public class AddDynamicPageFeed {
+  private static final int PAGE_SIZE = 1_000;
+
+  private static class AddDynamicPageFeedParams extends CodeSampleParams {
+
+    @Parameter(names = ArgumentNames.CUSTOMER_ID, required = true)
+    private long customerId;
+
+    @Parameter(names = ArgumentNames.CAMPAIGN_ID, required = true)
+    private long campaignId;
+
+    @Parameter(names = ArgumentNames.AD_GROUP_ID, required = true)
+    private long adGroupId;
+  }
+
+  private static class FeedDetails {
+    private String feedResourceName;
+    private long urlAttributeId;
+    private long labelAttributeId;
+
+    private FeedDetails(String feedResourceName, long urlAttributeId, long labelAttributeId) {
```

Comment:
I think it's a useful convenience because the feed attributes are returned as a list containing a name and ID. That means you would have to loop through that list and search for the "Page URL" and "Label" names in order to get the associated IDs in both the createFeedMapping and createFeedItems methods. Within each of those methods, you'd either have to loop through the attributes twice (once for each attribute) or create a map in each method to identify which ID belongs to which name. I think the current implementation is cleaner and easier.

Reply:
How about this instead?

1) Have `createFeed` return the Feed resource name.
2) Have `getFeed` take the Feed resource name as a parameter and return a `Map<DsaPageFeedCriterionField, FeedAttribute>`.
2) Have `createFeedMapping` take the Feed resource name and a `Map<DsaPageFeedCriterionField, FeedAttribute>` as parameters, and return a `Map<DsaPageFeedCriterionField, AttributeFieldMapping>'.
3) Have `createFeedItem` take the Feed resource name and a `Map<DsaPageFeedCriterionField, AttributeFieldMapping>' as parameters.

The benefit of this approach is that it uses API types and thereby helps users gain an understanding of what those types represent and contain. It also uses the `DsaPageCriterionField` enum values in place of names.

Reply:
SGTM, updated.